From 185178ef52d6f665a53c69f794daf8c2ec15b455 Mon Sep 17 00:00:00 2001
From: James Chapman <jachapma@redhat.com>
Date: Thu, 26 Feb 2026 12:55:47 +0000
Subject: [PATCH] Issue 7284 - Creating local password policy succeeds with
 incorrect passwordInHistory value (#7285)

Description:
attr_check_minmax used strtol(value, NULL, 0), which silently converted
invalid strings to 0, passing subsequent range checks.

Fix:
Add checks for NULL or empty values and uses strtol with endptr to
validate int input before range checks.

Fixes: https://github.com/389ds/389-ds-base/issues/7284

Reviewed by: @vashirov, @tbordaz (Thank you)
---
 ldap/servers/slapd/attr.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/ldap/servers/slapd/attr.c b/ldap/servers/slapd/attr.c
index f5ea22eb5..ca2eb62c3 100644
--- a/ldap/servers/slapd/attr.c
+++ b/ldap/servers/slapd/attr.c
@@ -932,8 +932,20 @@ attr_check_minmax(const char *attr_name, char *value, long minval, long maxval,
 {
     int retVal = LDAP_SUCCESS;
     long val;
+    char *endptr = NULL;
+
+    if (!value || *value == '\0') {
+        slapi_create_errormsg(errorbuf, ebuflen, "%s: attr value is NULL.", attr_name);
+        return LDAP_CONSTRAINT_VIOLATION;
+    }
+
+    errno = 0;
+    val = strtol(value, &endptr, 0);
+    if (endptr == value || *endptr != '\0' || errno != 0) {
+        slapi_create_errormsg(errorbuf, ebuflen, "%s: invalid value \"%s\".", attr_name, value);
+        return LDAP_CONSTRAINT_VIOLATION;
+    }
 
-    val = strtol(value, NULL, 0);
     if ((minval != -1 ? (val < minval ? 1 : 0) : 0) ||
         (maxval != -1 ? (val > maxval ? 1 : 0) : 0)) {
         slapi_create_errormsg(errorbuf, ebuflen, "%s: invalid value \"%s\".", attr_name, value);
-- 
2.53.0

